File: src/main/java/com/hazza/leetcode/tree/No110BalancedBinaryTree.java
package com.hazza.leetcode.tree;

import com.hazza.leetcode.datastructure.TreeNode;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/balanced-binary-tree/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 18-1-22
 * Time: 4:09 PM
 */
public class No110BalancedBinaryTree {

    // solution 1
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        int leftDepth = getDepth(root.left);
        int rightDepth = getDepth(root.right);

        return Math.abs(leftDepth - rightDepth) <= 1 && isBalanced(root.left) && isBalanced(root.right);
    }

    private int getDepth(TreeNode node) {
        if (node == null) return 0;
        return Math.max(getDepth(node.left), getDepth(node.right)) + 1;
    }

    // solution 2
    public boolean isBalanced1(TreeNode root) {
        return depthDFS(root) != -1;
    }

    private int depthDFS(TreeNode node) {
        if (node == null) return 0;

        int leftDepth = depthDFS(node.left);
        if (leftDepth == -1) return -1;
        int rightDepth = depthDFS(node.right);
        if (rightDepth == -1) return -1;

        if (Math.abs(leftDepth - rightDepth) > 1) return -1;
        return Math.max(leftDepth, rightDepth) + 1;
    }

}

File: src/main/python/leetcode_by_python/tree/No110_Balanced_Binary_Tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-09-19
# https://leetcode.com/problems/balanced-binary-tree/discuss/?currentPage=1&orderBy=most_votes&query=
from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    # solution 1
    def isBalanced1(self, root: TreeNode) -> bool:
        if not root:
            return True
        leftDepth = self.depth(root.left)
        rightDepth = self.depth(root.right)

        return abs(leftDepth - rightDepth) <= 1 and \
            self.isBalanced1(root.left) and \
            self.isBalanced1(root.right)

    def depth(self, node):
        if not node:
            return 0
        return max(self.depth(node.left), self.depth(node.right)) + 1

    # solution 2
    def isBalanced2(self, root: TreeNode) -> bool:
        return self.dfs(root) != -1

    def dfs(self, node):
        if not node:
            return 0
        leftDepth = self.dfs(node.left)
        if leftDepth == -1:
            return -1
        rightDepth = self.dfs(node.right)
        if rightDepth == -1:
            return -1
        if abs(leftDepth - rightDepth) > 1:
            return -1

        return max(leftDepth, rightDepth) + 1
